### PR TITLE
ddl: support alter algorithm INPLACE/INSTANT

### PR DESCRIPTION
**Pull request by winkyao**
### What problem does this PR solve? 

For #8428 , this is the first step to support `EXPLAIN DDL`(#8429).

MySQL support COPY/INPLACE/INSTANT/DEFAULT alter algorithm, you can get the details in https://dev.mysql.com/doc/refman/8.0/en/alter-table.html#alter-table-performance

In TiDB, we only support INPLACE/INSTANT, and the most algorithm of the alter operations is the `INSTANT`, which means 
> Operations only modify metadata in the data dictionary. No exclusive metadata locks are taken on the table during preparation and execution, and table data is unaffected, making operations instantaneous. Concurrent DML is permitted.

The add index using `INPLACE` algorithm, but not exactly the same with MySQL, the difference with MySQL is that TiDB never blocks the DML and never have a metadata lock on the table:

>Operations avoid copying table data but may rebuild the table in place. An exclusive metadata lock on the table may be taken briefly during the preparation and execution phases of the operation. Typically, concurrent DML is supported.

If user specifies the unsupported algorithm of the alter operation, a wanrning `ERROR 1846 (0A000): ALGORITHM=COPY is not supported. Reason: Cannot alter table by COPY. Try ALGORITHM=INSTANT.` will be returned:

```
mysql> create table t(a int);
Query OK, 0 rows affected (0.00 sec)

mysql> alter table t comment = 'test', ALGORITHM=COPY;
Query OK, 0 rows affected, 1 warning (0.00 sec)

mysql> show warnings;
+-------+------+---------------------------------------------------------------------------------------------+
| Level | Code | Message                                                                                     |
+-------+------+---------------------------------------------------------------------------------------------+
| Error | 1846 | ALGORITHM=COPY is not supported. Reason: Cannot alter table by COPY. Try ALGORITHM=INSTANT. |
+-------+------+---------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

Need to merge https://github.com/pingcap/parser/pull/93 first.

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to update the documentation
 - Need to be included in the release note

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8811)

@zimulala @crazycs520 @morgo  PTAL

### CODE REVIEW

**Comment by winkyao** [comment]
The circleci will be fixed after the https://github.com/pingcap/parser/pull/93 mereged.

File: ddl/ddl_algorithm_test.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
+// Copyright 2015 PingCAP, Inc.
```
Comment (shenli) [inline ddl/ddl_algorithm_test.go]: 2015 -> 2018

File: ddl/ddl_algorithm.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
+// Copyright 2013 The ql Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSES/QL-LICENSE file.
+
+// Copyright 2015 PingCAP, Inc.
```
Comment (shenli) [inline ddl/ddl_algorithm.go]: 2015 -> 2018

File: go.mod [hunk 3]
```diff
@@ -86,3 +86,5 @@ require (
 	sourcegraph.com/sourcegraph/appdash v0.0.0-20180531100431-4c381bd170b4
 	sourcegraph.com/sourcegraph/appdash-data v0.0.0-20151005221446-73f23eafcf67
 )
+
+replace github.com/pingcap/parser => github.com/winkyao/parser v0.0.0-20181221090435-0eaad9528419
```
Comment (shenli) [inline go.mod]: Should we remove this line?

**Comment by morgo** [comment]
@winkyao I spoke to MySQL users about the error on `ALGORITHM=COPY` behavior, and they reminded me that TiDB might be replicating from a MySQL system.  A warning is better behavior, since it won't break replication if an upstream MySQL server sends this DDL event.

**Comment by GregoryIan** [comment]
@winkyao what do you think？

**Comment by winkyao** [comment]
@morgo Well, return warning is also acceptable for me. I'll change this PR later.

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -1407,10 +1407,16 @@ func getCharsetAndCollateInTableOption(startIdx int, options []*ast.TableOption)
 	return
 }
 
-func (d *ddl) AlterTable(ctx sessionctx.Context, ident ast.Ident, specs []*ast.AlterTableSpec) (err error) {
-	// Only handle valid specs.
+// resolveAlterTableSpec resolve alter table algorithm and remove ignore table spec in specs.
```
Comment (zimulala) [inline ddl/ddl_api.go]: s/resolve/resolves
s/remove/removes

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -1407,10 +1407,16 @@ func getCharsetAndCollateInTableOption(startIdx int, options []*ast.TableOption)
...
+// returns valied specs, and the occured error.
```
Comment (zimulala) [inline ddl/ddl_api.go]: s/returns/Return

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -1407,10 +1407,16 @@ func getCharsetAndCollateInTableOption(startIdx int, options []*ast.TableOption)
...
+func resolveAlterTableSpec(ctx sessionctx.Context, specs []*ast.AlterTableSpec) ([]*ast.AlterTableSpec, error) {
 	validSpecs := make([]*ast.AlterTableSpec, 0, len(specs))
+	algorithm := ast.AlterAlgorithmDefault
 	for _, spec := range specs {
+		if spec.Tp == ast.AlterTableAlgorithm {
+			// find the last AlterTableAlgorithm.
```
Comment (zimulala) [inline ddl/ddl_api.go]: s/find/Find

File: ddl/db_integration_test.go [hunk 7]
```diff
@@ -1210,3 +1210,75 @@ func (s *testIntegrationSuite) TestAlterColumn(c *C) {
 	_, err = s.tk.Exec("alter table mc modify column a bigint auto_increment") // Adds auto_increment should throw error
 	c.Assert(err, NotNil)
 }
+
+func (s *testIntegrationSuite) assertWarningExec(c *C, sql string) {
```
Comment (zimulala) [inline ddl/db_integration_test.go]: I think adding the warning message as an argument can be beneficial for calls to other functions.

**Comment by winkyao** [comment]
@zimulala Done, PTAL

**Pull request merged**